### ISSUE TITLE
InnerAssignment: false positives on for and do-while loops

### ISSUE THREAD

**Issue by strkkk** [checkstyle/checkstyle#6915]
_Originally posted by @romani in https://github.com/checkstyle/checkstyle/pull/6886_

from https://checkstyle.org/config_coding.html#InnerAssignment
> Rationale: With the exception of for iterators and assignment in while idiom

```
$ cat config.xml 
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
  "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
  "https://checkstyle.org/dtds/configuration_1_3.dtd">
<module name="Checker">
  <module name="TreeWalker">
    <module name="InnerAssignment"/>
  </module>
</module>

$ cat Test.java 
import java.io.Reader;
import java.io.IOException;

class Test {
  public void foo(Reader reader) throws IOException {
    int value;
    while ((value = reader.read()) != 0) { // OK, no violation is expected
      // process
    }

    do {
      // process
    } while ((value = reader.read()) != 0);  // violation, but no expected

    for (;(value = reader.read()) != 0;) {  // violation, but no expected
      // process
    }
  }
}

$ java -jar checkstyle-8.23-all.jar -c config.xml Test.java
Starting audit...
[ERROR] Test.java:13:21: Inner assignments should be avoided. [InnerAssignment]
[ERROR] Test.java:15:18: Inner assignments should be avoided. [InnerAssignment]
Audit done.
Checkstyle ends with 2 errors.
```

Expected:  no violations in all cases of loop. 

ATTENTION: After fix, doc should be  updated to name idion it as `Loop-idiom` 
and make tests for FOR/do_while too.

**Comment by romani** [comment]
@rnveach , please review and approve

**Comment by DXTkastb** [comment]
I'm on it.

**Comment by DXTkastb** [comment]
> Expected: no violations in all cases of loop.

@strkkk  just a small doubt , do we need to skip checking for `for` and `do-while` loops (no violation whatsoever)
Or
 we have to keep in mind the situation:  `for (;u=true;)`
because in this case Check throws violation too.

**Comment by strkkk** [comment]
@DXTkastb it should be skipped for expressions in loops condition

**Comment by romani** [comment]
fix is merged
